### PR TITLE
feat: rpad support column for second arg instead of just literal

### LINKED ISSUE #2096: rpad expression panics if length input is not a literal value
### Describe the bug

The rpad expression panics if the `length` input is not a literal value.

For example, `rpad(string_column, 10)` works, but `rpad(string_column, int_column)` panics.

This is probably not a common usage, but we should fall back to Spark rather than panic.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?

Closes #2096 

Implement comet native logic to support rpad(column, column) API in Spark . Currently comet only supports rpad(column, int)

## What changes are included in this PR?

PR to implement native code to support rpad(col, int)

## How are these changes tested?

Unit testing in cometSuite

### CODE REVIEW

**Comment by coderfender** [comment]
@andygrove  , The issue is with implementation of `rpad` to only support col,int signature . Rather than reverting to native spark code, I went ahead and implemented native code for `col,col` input (and added a test case in `CometExpressionSuite` .  Please take a look at the changes and let me know your thoughts on the changes

**Comment by coderfender** [comment]
Thank you for the review @comphead . Moving expressions to datafusion-spark create is indeed the goal once this change is merged into main

**Comment by coderfender** [comment]
@andygrove  , @comphead  could you please review the code whenever you get a chance ? Thank you very much

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 1]
```diff
@@ -71,44 +101,101 @@ fn spark_read_side_padding2(
...
-                            .unwrap_or(string.len());
-                        builder.append_value(&string[..idx]);
-                    } else {
-                        builder.append_value(string);
+            for string in string_array.iter() {
+                match string {
+                    Some(string) => {
+                        // It looks Spark's UTF8String is closer to chars rather than graphemes
+                        // https://stackoverflow.com/a/46290728
+                        let char_len = string.chars().count();
```
Comment (mbutrovich) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Let's abstract out the duplicate code so if we change the logic it's a smaller blast radius.

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 2]
```diff
@@ -71,44 +101,101 @@ fn spark_read_side_padding2(
...
     truncate: bool,
+    rpad_argument: RPadArgument,
 ) -> Result<ColumnarValue, DataFusionError> {
     let string_array = as_generic_string_array::<T>(array)?;
-    let length = 0.max(length) as usize;
-    let space_string = " ".repeat(length);
+    match rpad_argument {
+        RPadArgument::ColArray(array_int) => {
+            let int_pad_array = array_int.as_primitive::<Int32Type>();
+            let mut str_pad_value_map = HashMap::new();
```
Comment (mbutrovich) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: I might be missing something obvious: why do we need a `HashMap`? It doubles the memory usage of this operator. If we iterate through the arrays to build the `HashMap`, why can't we iterate through the arrays with the builder below (i.e. `zip` through them or something). This approach would also have the benefit of preserving input order (though I'm not sure that's required).

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs (hunk 1 above)
Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Thank you for the feedback @mbutrovich . I moved the string manipulation logic to a separate function and leveraged it in both the use cases.

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs (hunk 2 above)
Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Thats a great suggestion @mbutrovich  . I went ahead and  removed hashmap creation and changed logic iterate through the arrays together .

**Comment by coderfender** [comment]
@mbutrovich  , seems like a test failed due to a perhaps transient Spark env issue. Could you rerun the failed check whenever you get a chance please ?

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 3]
```diff
@@ -42,18 +41,48 @@ fn spark_read_side_padding2(
...
+                ))),
+            }
+        }
+        [ColumnarValue::Array(array), ColumnarValue::Array(array_int)] => {
+            let rpad_arg = RPadArgument::ColArray(Arc::clone(array_int));
+            match array.data_type() {
+                DataType::Utf8 => {
+                    spark_read_side_padding_internal::<i32>(array, truncate, rpad_arg)
+                }
+                DataType::LargeUtf8 => {
```
Comment (mbutrovich) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: When we bring this to DataFusion we will need to support `Utf8View`. We can't really test that in Comet without a unit test in the file, but something to prepare for.

Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Thank you for sharing this @mbutrovich I will update this info in the github issue I plan to create to port these changes to data fusion crate

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 4]
```diff
@@ -71,44 +100,78 @@ fn spark_read_side_padding2(
     }
 }
 
+enum RPadArgument {
```
Comment (mbutrovich) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Why do we need a new `enum` type instead of relying on `ColumnarValue` when can already represent a scalar or array?

Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Thank you . This is great suggestion  and I went ahead and leveraged `ColumnarValue` to fork to the right logic

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 5]
```diff
@@ -43,17 +42,69 @@ fn spark_read_side_padding2(
...
+                    array,
+                    truncate,
+                    ColumnarValue::Scalar(ScalarValue::Int32(Some(*length))),
+                ),
+                DataType::LargeUtf8 => spark_read_side_padding_internal::<i64>(
+                    array,
+                    truncate,
+                    ColumnarValue::Scalar(ScalarValue::Int32(Some(*length))),
+                ),
+                // Dictionary support required for SPARK-48498
```
Comment (mbutrovich) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs:55]: https://github.com/apache/spark/pull/46832

This seems related to padding. How does this affect dictionary encoded columns?

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 6]
```diff
@@ -43,17 +42,69 @@ fn spark_read_side_padding2(
...
+                    array,
+                    truncate,
+                    ColumnarValue::Array(Arc::<dyn arrow::array::Array>::clone(array_int)),
+                ),
+                DataType::LargeUtf8 => spark_read_side_padding_internal::<i64>(
+                    array,
+                    truncate,
+                    ColumnarValue::Array(Arc::<dyn arrow::array::Array>::clone(array_int)),
+                ),
                 // Dictionary support required for SPARK-48498
```
Comment (mbutrovich) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Same question.

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 7]
```diff
@@ -322,6 +322,16 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       checkSparkAnswer("SELECT try_add(_1, _2) FROM tbl")
     }
   }
+  test("fix_rpad") {
```
Comment (comphead) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: can we get the meaningful test name? what exactly fix is tested

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 8]
```diff
@@ -73,42 +124,71 @@ fn spark_read_side_padding2(
...
+                    _ => builder.append_null(),
                 }
             }
-            _ => builder.append_null(),
+            Ok(ColumnarValue::Array(Arc::new(builder.finish())))
+        }
+    }
+}
+
+fn add_padding_string(string: String, length: usize, truncate: bool) -> String {
```
Comment (comphead) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: perhaps we can think of impl like 

```
fn add_padding_string(input: String, length: usize, truncate: bool) -> String {
    let char_len = input.chars().count();

    if char_len >= length {
        if truncate {
            // Take the first `length` chars safely
            input.chars().take(length).collect()
        } else {
            input
        }
    } else {
        // Pad with only the needed spaces
        let padding = " ".repeat(length - char_len);
        input + &padding
    }
}
```

so we don't allocate spaces if its not needed
no unwrap

Comment (comphead) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: refering string by index, is it unicode safe? 🤔

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala (hunk 7 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: Sure . Thank you for the review. I will update the test name to add more context

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs (hunk 5 above)
Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs:55]: Great catch! My understanding is that dictionary support ensures SQL-compliant CHAR type literals, which always have a fixed length. Therefore, my support for the array argument is obsolete.

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs (hunk 6 above)
Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: Great catch! My understanding is that dictionary support ensures SQL-compliant CHAR type literals, which always have a fixed length (This change already existed by the time I picked up this issue). Therefore, my support for the array argument is obsolete.

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs (hunk 8 above)
Comment (coderfender) [reply native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs]: This is a great suggestion. My goal for now was to keep the original implementation intact and not introduce changes which directly doesn't solve the issue

**Review by comphead — APPROVED** [review]
Thanks @coderfender its LGTM
Please add a test for unicode string to see if there is an issue, if it is we need to comment the test to be fixed in the future, and also we probably need to document this limitations

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 9]
```diff
@@ -73,42 +101,71 @@ fn spark_read_side_padding2(
...
+
+            let mut builder = GenericStringBuilder::<T>::with_capacity(
+                string_array.len(),
+                string_array.len() * length,
+            );
+
+            for string in string_array.iter() {
+                match string {
+                    Some(string) => builder.append_value(add_padding_string(
+                        string.parse().unwrap(),
```
Comment (comphead) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs:140]: its good to avoid unwraps and return Err instead

File: native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs [hunk 10]
```diff
@@ -73,42 +101,71 @@ fn spark_read_side_padding2(
...
-                    }
-                } else {
-                    // write_str updates only the value buffer, not null nor offset buffer
-                    // This is convenient for concatenating str(s)
-                    builder.write_str(string)?;
-                    builder.append_value(&space_string[char_len..]);
+            for (string, length) in string_array.iter().zip(int_pad_array) {
+                match string {
+                    Some(string) => builder.append_value(add_padding_string(
+                        string.parse().unwrap(),
```
Comment (comphead) [inline native/spark-expr/src/static_invoke/char_varchar_utils/read_side_padding.rs:120]: ditto

**Comment by coderfender** [comment]
@comphead thank you for the approval . I did make changes to add UTF string test case and also remove unsafe unwrap method and handled error through match statement . Please take a look whenever you get a chance . Thank you

**Comment by coderfender** [comment]
Thank you for the approval @mbutrovich  , @comphead  . I will be utilizing the same code base with perhaps some wrapper function to  implemnent support for `lpad` expression .

**Comment by coderfender** [comment]
Thank you for merging this to main @mbutrovich  , @comphead . I will be leveraging the same code above (and some boiler plate) to implement lpad expression support

**Comment by coderfender** [comment]
PR for lpad support : https://github.com/apache/datafusion-comet/pull/2102
